cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -630,24 +650,28 @@ class AsofJoinNode : public ExecNode {
       for (int i = 0; i < input_schema->num_fields(); ++i) {
         const auto field = input_schema->field(i);
         if (field->name() == on_field_name) {
-          if (kSupportedOnTypes_.find(field->type()) == kSupportedOnTypes_.end()) {
-            return Status::Invalid("Unsupported type for on key: ", field->name());
+          // Equals must be used when checking for timestamp types.
+          if (!field->type()->Equals(TimestampType(TimeUnit::NANO, "UTC")) &&
```

Comment:
Can you put TimestampType in kSupportedOnTypes_

Reply:
e.g, even though I have added `TimestampType` to `kSupportedOnTypes_`, it has no effect. I looked for a more generic way to detect a `TimestampType`, like `arrow::istimestamp_type` in `src/arrow/type_traits.h` but I'm not sure how it works with the templating.